test/cbmc/sources/mqtt_agent_cbmc_state.c

```diff
@@ -0,0 +1,139 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file mqtt_agent_cbmc_state.c
+ * @brief Implements the functions defined in mqtt_agent_cbmc_state.h.
+ */
+#include <stdint.h>
+#include <stdlib.h>
+#include "mqtt_agent.h"
+#include "mqtt_agent_cbmc_state.h"
+#include "network_interface_stubs.h"
+#include "get_time_stub.h"
+#include "incoming_publish_callback_stub.h"
+#include "agent_message_stubs.h"
+#include "agent_command_pool_stubs.h"
+
+
+MQTTFixedBuffer_t * allocateMqttFixedBuffer( MQTTFixedBuffer_t * pFixedBuffer )
+{
+    if( pFixedBuffer == NULL )
+    {
+        pFixedBuffer = malloc( sizeof( MQTTFixedBuffer_t ) );
+    }
+
+    if( pFixedBuffer != NULL )
+    {
+        __CPROVER_assume( pFixedBuffer->size < CBMC_MAX_OBJECT_SIZE );
+        pFixedBuffer->pBuffer = malloc( pFixedBuffer->size );
+    }
+
+    return pFixedBuffer;
+}
+
+bool isValidMqttFixedBuffer( const MQTTFixedBuffer_t * pFixedBuffer )
+{
+    bool isValid = true;
+
+    if( pFixedBuffer != NULL )
+    {
+        isValid = pFixedBuffer->size < CBMC_MAX_OBJECT_SIZE;
+    }
+
+    return isValid;
+}
+
+MQTTAgentContext_t * allocateMqttAgentContext( MQTTAgentContext_t * pContext )
+{
+    TransportInterface_t * pTransportInterface;
+    MQTTFixedBuffer_t * pNetworkBuffer;
+    AgentMessageInterface_t * pMessageInterface;
+    MQTTStatus_t status = MQTTSuccess;
+    void * pIncomingCallbackContext;
+
+    if( pContext == NULL )
+    {
+        pContext = malloc( sizeof( MQTTAgentContext_t ) );
+    }
+
+    pTransportInterface = malloc( sizeof( TransportInterface_t ) );
+
+    if( pTransportInterface != NULL )
+    {
+        /* The possibility that recv and send callbacks are NULL is tested in the
+         * MQTT_Init proof. MQTT_Init is required to be called before any other
+         * function in core_mqtt.h. */
+        pTransportInterface->recv = NetworkInterfaceReceiveStub;
+        pTransportInterface->send = NetworkInterfaceSendStub;
+    }
+
+    pNetworkBuffer = allocateMqttFixedBuffer( NULL );
+    __CPROVER_assume( isValidMqttFixedBuffer( pNetworkBuffer ) );
+
+    pMessageInterface = malloc( sizeof( AgentMessageInterface_t ) );
+
+    if( pMessageInterface != NULL )
+    {
+        pMessageInterface->send = AgentMessageSendStub;
+        pMessageInterface->recv = AgentMessageRecvStub;
+        pMessageInterface->getCommand = AgentGetCommandStub;
+        pMessageInterface->releaseCommand = Agent_ReleaseCommand;
+    }
```

Comment:
This is fine, but I remember Mark preferred having an `isValidMessageInterface` where we say `__CPROVER_assume(pMessageInterface->releaseCommand == Agent_ReleaseCommand)`. Same for other assignments inside this block.

Reply:
Yea. This is another way to handle it. However from a speed of the proof I prefer the assignment here. Do you know of any benefits of using __CPROVER_assume instead?

Reply:
That would be a good question to ask in the CBMC channel.

Reply:
Yea checking with Mark from ARG team. now. As discussed offline, if he points out about a benefit in that approach, I can create a follow up PR to address that. Thanks

Reply:
Generally speaking, the thing I care about most is the set of assumptions that the proof depends on.  I want to know that all assumptions are listed in one place (or in a well-defined and documented set of places).  So I want the proof harness to list the collection of proof assumptions as a list of assumptions.  You can, of course, model assumptions like initializations in the form of assignments in the harness, but that makes the assumptions implicit and harder to find with a scan of the harness.  It is rare, I think, that using assume(a=b) in place of assignment a=b slows down the proof checking.

Reply:
Thanks Mark. I will make a follow up change to update the assignments into assumptions